src/util/http/websocket/QueryHub.cpp

```diff
@@ -0,0 +1,53 @@
+//   Copyright 2023, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#include "util/http/websocket/QueryHub.h"
+
+#include "util/Asio.h"
+
+namespace ad_utility::websocket {
+net::awaitable<std::shared_ptr<QueryToSocketDistributor>>
+QueryHub::createOrAcquireDistributorInternal(QueryId queryId) {
+  co_await net::dispatch(globalStrand_, net::use_awaitable);
+  if (socketDistributors_.contains(queryId)) {
+    if (auto ptr = socketDistributors_.at(queryId).pointer_.lock()) {
+      co_return ptr;
+    }
+    // There's a scenario where the object has already been destructed, but not
+    // yet removed from the list. In this case remove this here and proceed
+    // to build a new instance. This is safe because the id is checked for
+    // equality before removal.
+    socketDistributors_.erase(queryId);
+  }
+
+  auto id = counter++;
+  auto coroutine = [](auto strand, auto& distributors, auto queryId,
```

Comment:
Should be called cleanupAction or `unregisterDistributorFromHub` or something and be commented.